src/Data/Text/Encoding.hs

```diff
@@ -187,32 +189,295 @@ decodeLatin1 bs = withBS bs $ \fp len -> runST $ do
 foreign import ccall unsafe "_hs_text_is_ascii" c_is_ascii
     :: Ptr Word8 -> Ptr Word8 -> IO CSize
 
-isValidBS :: ByteString -> Bool
+data Progression
+  = WriteAndAdvance Char Int
+  | NeedMore
+  | Invalid
+
+decodeChunks ::
+#if defined(ASSERTS)
+  HasCallStack =>
+#endif
+  (Bits w, Num w, Storable w)
+  => w
+  -> ( Int
+    -> ByteString
+    -> A.MArray s
+    -> Int
+    -> Int
+    -> Int
+    -> ST s (Maybe (
+        (A.MArray s -> Int -> Int -> Int -> ST s (DecodeResult Text ByteString w))
+        -> ST s (DecodeResult Text ByteString w))
+      )
+    )
+  -> ((Int -> Word8) -> Int -> Int -> Progression)
+  -> ByteString
+  -> ByteString
+  -> ST s (DecodeResult Text ByteString w)
+decodeChunks w bulkDecodeF incrementalDecodeF bs1@(B.length -> len1) bs2@(B.length -> len2) = do
```

Comment:
Is there no way to simplify this? as far as `decodeChunks` is concerned, `bulkDecodeF` parses a prefix of the source and copies it into the destination. This overlaps with the duties of `incrementalDecodeF`, which can be seen as parsing a one-character prefix. Can't there be a function that turns a `incrementalDecodeF` into a `bulkDecodeF`, so that `decodeChunks` then only takes a single function parameter?

Reply:
See my response to other concern you raised below.